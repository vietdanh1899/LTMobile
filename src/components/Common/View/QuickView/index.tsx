import React, { PureComponent } from 'react';
import {
  View,
  ImageBackgroundProps,
  ImageBackground,
  StyleSheet,
  ViewProps,
  TouchableOpacityProps,
  TouchableOpacity,
  ScrollView,
  ScrollViewProps,
  TouchableWithoutFeedback,
  Keyboard,
  GestureResponderEvent,
} from 'react-native';
import {
  roundedBorderRadius,
  shadowViewLight,
} from '@themes/ThemeComponent/Common/CommonProps';
import _ from 'lodash';

export interface QuickViewProps
  extends ViewProps,
  TouchableOpacityProps,
  ScrollViewProps {
  width?: number | string;
  height?: number | string;
  margin?: number;
  marginTop?: number;
  marginBottom?: number;
  marginLeft?: number;
  marginRight?: number;
  marginHorizontal?: number;
  marginVertical?: number;
  padding?: number;
  paddingTop?: number;
  paddingBottom?: number;
  paddingLeft?: number;
  paddingRight?: number;
  paddingHorizontal?: number;
  paddingVertical?: number;
  borderRadius?: number;
  borderTopLeftRadius?: number;
  borderTopRightRadius?: number;
  borderBottomLeftRadius?: number;
  borderBottomRightRadius?: number;
  position?: 'absolute' | 'relative';
  top?: number;
  bottom?: number;
  left?: number;
  right?: number;
  center?: boolean;
  horizontalCenter?: boolean;
  verticalCenter?: boolean;
  style?: any;
  children?: any;
  row?: boolean;
  column?: boolean;
  rowReverse?: boolean;
  columnReverse?: boolean;
  justifyContent?:
  | 'center'
  | 'flex-start'
  | 'flex-end'
  | 'space-around'
  | 'space-between';
  alignSelf?:
  | 'auto'
  | 'center'
  | 'flex-start'
  | 'flex-end'
  | 'stretch'
  | 'baseline';
  alignItems?:
  | 'auto'
  | 'center'
  | 'flex-start'
  | 'flex-end'
  | 'stretch'
  | 'baseline';
  backgroundColor?: string;
  flex?: number;
  backgroundImage?: ImageBackgroundProps;
  backgroundImageStyle?: any;
  sharp?: boolean;
  rounded?: boolean;
  circle?: boolean;
  scroll?: boolean;
  touchableComponent?: 'TouchableOpacity' | 'TouchableWithoutFeedback';
  dismissKeyboard?: boolean;
  shadow?: boolean;
  testID?: string;
}

class QuickView extends PureComponent<QuickViewProps> {
  static defaultProps = {
    column: true,
    touchableComponent: 'TouchableOpacity',
  };

  render() {
    const {
      width: widthProp,
      height: heightProp,
      margin,
      marginTop,
      marginBottom,
      marginLeft,
      marginRight,
      marginHorizontal,
      marginVertical,
      padding,
      paddingTop,
      paddingBottom,
      paddingLeft,
      paddingRight,
      paddingHorizontal,
      paddingVertical,
      borderRadius: borderRadiusProp,
      borderTopLeftRadius,
      borderTopRightRadius,
      borderBottomLeftRadius,
      borderBottomRightRadius,
      position,
      top,
      bottom,
      left,
      right,
      center,
      horizontalCenter,
      verticalCenter,
      style,
      children,
      row: rowProp,
      column: columnProp,
      rowReverse: rowReverseProp,
      columnReverse: columnReverseProp,
      justifyContent,
      alignSelf,
      alignItems,
      backgroundColor,
      flex,
      backgroundImage,
      backgroundImageStyle,
      sharp: sharpProp,
      rounded: roundedProp,
      circle: circleProp,
      scroll,
      touchableComponent: touchableComponentProp,
      dismissKeyboard,
      shadow,
      onPress: onPressProp,
      ...otherProps
    } = this.props;

    /**
     * containerStyle
     */
    let row = rowProp;
    let rowReverse = rowReverseProp;
    let column = columnProp;
    let columnReverse = columnReverseProp;
    if (row || rowReverse) {
      column = false;
      columnReverse = false;
    }
    if (column || columnReverse) {
      row = false;
      rowReverse = false;
    }

    /**
     * sharp, rounded, circle
     */
    let width = widthProp;
    let height = heightProp;
    let borderRadius: any = 0;
    let sharp = sharpProp;
    let rounded = roundedProp;
    let circle = circleProp;
    if (borderRadiusProp) {
      borderRadius = borderRadiusProp;
    } else {
      if (circle) {
        rounded = false;
        sharp = false;
        const minDimension = _.min([width, height]) || 50;
        width = minDimension;
        height = minDimension;
        borderRadius = minDimension;
      }
      if (rounded) {
        borderRadius = roundedBorderRadius;
      }
      if (sharp) {
        rounded = false;
        circle = false;
        borderRadius = 0;
      }
    }

    /**
     * containerStyle
     */
    const containerStyle = StyleSheet.flatten([
      {
        width,
        height,
        margin,
        marginTop,
        marginBottom,
        marginLeft,
        marginRight,
        marginHorizontal,
        marginVertical,
        padding,
        paddingTop,
        paddingLeft,
        paddingRight,
        paddingHorizontal,
        paddingBottom,
        paddingVertical,
        borderRadius,
        borderTopLeftRadius,
        borderTopRightRadius,
        borderBottomLeftRadius,
        borderBottomRightRadius,
        position,
        top,
        bottom,
        left,
        right,
        justifyContent,
        alignSelf,
        alignItems,
        backgroundColor,
        flex,
      },
      shadow && { paddingBottom: paddingBottom || 3 },
      shadow && shadowViewLight,
      center && {
        alignSelf: 'center',
        alignItems: 'center',
        justifyContent: 'center',
      },
      ((horizontalCenter && !(row || rowReverse))
        || (verticalCenter && (row || rowReverse))) && {
        alignItems: 'center',
      },
      ((horizontalCenter && (row || rowReverse))
        || (verticalCenter && !(row || rowReverse))) && {
        justifyContent: 'center',
      },
      row && { flexDirection: 'row' },
      column && { flexDirection: 'column' },
      rowReverse && { flexDirection: 'row-reverse' },
      columnReverse && { flexDirection: 'column-reverse' },
      style,
    ]);

    let onPress: any = onPressProp;
    let touchableComponent = touchableComponentProp;
    if (dismissKeyboard) {
      onPress = (event: GestureResponderEvent) => {
        Keyboard.dismiss();
        if (onPressProp) onPressProp(event);
      };
      touchableComponent = 'TouchableWithoutFeedback';
    }

    let Component: any;
    if (onPress) {
      switch (touchableComponent) {
        case 'TouchableWithoutFeedback':
          Component = TouchableWithoutFeedback;
          break;
        default:
          Component = TouchableOpacity;
          break;
      }
    } else if (scroll) {
      Component = ScrollView;
    } else {
      Component = View;
    }

    if (backgroundImage) {
      const backgroundStyle: any = [
        {
          flex: 1,
          resizeMode: 'cover',
        },
        backgroundImageStyle,
      ];
      return (
        <ImageBackground style={backgroundStyle} {...backgroundImage}>
          <Component {...otherProps} onPress={onPress} style={containerStyle}>
            {children}
          </Component>
        </ImageBackground>
      );
    }
    return (
      <Component {...otherProps} style={containerStyle} onPress={onPress}>
        {children}
      </Component>
    );
  }
}

export default QuickView;
