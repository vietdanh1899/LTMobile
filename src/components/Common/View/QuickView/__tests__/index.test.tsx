import React from 'react';
import {
  getStyle, testStyleProps, basicTest, EnumComponent, getComponent,
} from '@utils/testHelper';
import QuickView from '..';

const target = <QuickView />;
const type = EnumComponent.SINGLE;

describe('QuickView Component', () => {
  basicTest(target, type);

  it('should apply all style-related props', () => {
    const props = {
      width: 10,
      height: 10,
      margin: 10,
      marginVertical: 10,
      marginHorizontal: 10,
      marginTop: 10,
      marginBottom: 10,
      marginLeft: 10,
      marginRight: 10,
      padding: 10,
      paddingVertical: 10,
      paddingHorizontal: 10,
      paddingTop: 10,
      paddingBottom: 10,
      paddingLeft: 10,
      paddingRight: 10,
      borderRadius: 10,
      borderTopLeftRadius: 10,
      borderTopRightRadius: 10,
      borderBottomLeftRadius: 10,
      borderBottomRightRadius: 10,
      position: 'absolute',
      top: 10,
      bottom: 10,
      left: 10,
      right: 10,
      justifyContent: 'flex-start',
      alignSelf: 'flex-start',
      alignItems: 'flex-start',
      backgroundColor: '#FFFFFF',
      flex: 1,
    };
    testStyleProps(target, props, type);
  });

  it('should apply all custom props', () => {
    /**
     * center
     */
    const testStyle = getStyle(target, { center: true }, type);
    expect(testStyle.alignSelf).toBe('center');
    expect(testStyle.alignItems).toBe('center');
    expect(testStyle.justifyContent).toBe('center');

    /**
     * horizontalCenter
     */
    const testStyle2 = getStyle(target, { horizontalCenter: true }, type);
    expect(testStyle2.alignItems).toBe('center');
    expect(testStyle2.justifyContent).not.toBe('center');

    const testStyle3 = getStyle(target, { horizontalCenter: true, row: true }, type);
    expect(testStyle3.justifyContent).toBe('center');
    expect(testStyle3.alignItems).not.toBe('center');

    /**
     * verticalCenter
     */
    const testStyle4 = getStyle(target, { verticalCenter: true }, type);
    expect(testStyle4.alignItems).not.toBe('center');
    expect(testStyle4.justifyContent).toBe('center');

    const testStyle5 = getStyle(target, { verticalCenter: true, row: true }, type);
    expect(testStyle5.justifyContent).not.toBe('center');
    expect(testStyle5.alignItems).toBe('center');

    /**
     * TouchableOpacity || View
     */
    const onPress = jest.fn();
    const props1 = getComponent(target, { onPress }, type);
    expect(props1.name()).toBe('ForwardRef');

    const props2 = getComponent(target, {}, type);
    expect(props2.name()).toBe('View');
  });
});
