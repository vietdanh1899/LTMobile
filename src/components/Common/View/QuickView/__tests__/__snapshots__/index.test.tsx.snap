// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuickView Component should render without issues 1`] = `
<View
  style={
    Object {
      "alignItems": undefined,
      "alignSelf": undefined,
      "backgroundColor": undefined,
      "borderBottomLeftRadius": undefined,
      "borderBottomRightRadius": undefined,
      "borderRadius": 0,
      "borderTopLeftRadius": undefined,
      "borderTopRightRadius": undefined,
      "bottom": undefined,
      "flex": undefined,
      "flexDirection": "column",
      "height": undefined,
      "justifyContent": undefined,
      "left": undefined,
      "margin": undefined,
      "marginBottom": undefined,
      "marginHorizontal": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "marginVertical": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "paddingVertical": undefined,
      "position": undefined,
      "right": undefined,
      "top": undefined,
      "width": undefined,
    }
  }
/>
`;
